Connect with a seeded database

davidomarf opened the pull request:
This PR updates the `database` package to use a global `DBConn` variable instead of returning the db connection in the `GetConnection` function. 

This allows all the packages in `racegex` to import `database` and have access to `database.DBConn` without needing to be passing the connection around between functions.

---

The seeds are also created, and to demonstrate some usage, a new Model was added. 

For the seeds to work, these changes also do AutoMigration using `gorm`. 

Because we don't have users, and all the content for quite a good time will be all hardcoded, and the structs are very prone to change, I added a `DropTableIfExists(table)` before the migration to avoid having garbage columns. This also avoids problems when seeding repeated items.

davidomarf committed:
Create a global DBConn instead of returning the connection

davidomarf committed:
Create the models package

davidomarf committed:
Create the Tables for the models in the project using gorm.AutoMigrate

davidomarf committed:
Create the Seeds package and use it to seed the database

davidomarf committed:
Use the database.DBConn to obtain entries instead of faking it

davidomarf committed:
Add new endpoint controllers for lessons and add it in main

davidomarf committed:
Bump required packages

davidomarf committed:
Remove `primary_key` from JSON encoding in the structs

This was a flag for the gorm, but the ID fields are always the default
`primary_key`.

Please see: https://gorm.io/docs/conventions.html#ID-as-Primary-Key

davidomarf committed:
Use log.Panic instead of log.Fatal to make sure deferred functions run

`log.Fatal` terminates the flow immediately, not allowing deferred
funtions to run.

`log.Panic` just makes the program to start panicking.

From the blog:

"""
Panic is a built-in function that stops the ordinary flow of control
and begins panicking. When the function F calls panic, execution of F
stops, any deferred functions in F are executed normally, and then F
returns to its caller. To the caller, F then behaves like a call to
panic. The process continues up the stack until all functions in the
current goroutine have returned, at which point the program crashes.
Panics can be initiated by invoking panic directly. They can also be
caused by runtime errors, such as out-of-bounds array accesses.
"""

For more, please read: https://blog.golang.org/defer-panic-and-recover

davidomarf committed:
Update CHANGELOG

MikeG96 commented on main.go:
```diff
@@ -9,15 +9,28 @@ import (
+	// This creates the database.DBConn that is accessible by all packages
+	database.GetConnection()
+	// Create the Tables for the racegex/models
+	database.Migrate()
```
Creo que falta definir cuando debe de hacerlo y cuando no, algo más para saber en que environment estamos, en el archivo de configuración podria agregarse eso

davidomarf replied:
¿Tienes alguna sugerencia para manejar ambientes? 

Por ahora pensé que convenía migrar siempre. Ahorita ninguna estructura es permanente, ni bien definida. Van a seguir cambiando. 

Cuando lleguemos a algo estable podríamos cambiar el comportamiento por defecto. ¿Usamos alguna bandera?

MikeG96 replied:
```
{
  "datastore": {
    "address": "localhost",
    "port": "5432",
    "user": "golang",
    "password": "golang",
    "database": "racegex",
    "migrate": true | false, 
  }
}
```

Agregaria migrate a nuestra configuración local y así ya siempre se sabe que hay que hacer, sin embargo hay que agregar la rutina de borrado de base cuando este como true para que cada que se inicie el servidor dropear la base y volverla a generar

davidomarf committed:
Add a "migrate" and "drop" prop into the DB Configuration

This should act as a flag to decide whether the models should be
migrated or the database dropped when starting the server.

davidomarf committed:
Handle the conditional flags `migrate` and `drop`

The `drop` flag runs everytime a new migration is going to be made. So
there **must** be a model to migrate to try to drop it. Models that get
deleted won't get dropped in the database, since there's no reference to
it.
